### ISSUE TITLE
Tag parser should support quoting

### ISSUE THREAD

**Issue by tw-bert** [RedisLabsModules/RediSearch#259]
When querying for tags (i.e. `FT.SEARCH idxFoo @fld:{my-special_underscore-tag}` ), currently in-depth knowledge of character escaping is required for the end user in certain scenario's. This knowledge is __not__ required at `FT.ADD`, only at `FT.SEARCH`.

@dvirsky Suggested at the [RediSearch Mailgroup](https://groups.google.com/forum/#!topic/redisearch/x6vDoddCaGU) that enabling the user to quote queried tags is probably the best fix for this, taking [Ragel](https://github.com/bnoordhuis/ragel) performance into account. A side effect of this implementation will be that quotes are no valid tag delimiters anymore, which seems reasonable. 

Example: `FT.SEARCH idxFoo @fld:{"my-special_underscore-tag"}`

Note that quotes and double quotes __inside__ a tag can still be supported. To search for those situations, quotes and double quotes should be escaped by the client.

Example: `FT.SEARCH idxFoo @fld:{"with \"quotes\" embedded..."}`

Side issue: Tag searches of empty tags are not yet supported, but the [suggested](https://groups.google.com/forum/#!topic/redisearch/Tj-DOCRSeRQ) solution is using an underscore as the tag field value, when storing the document ( `FT.ADD` ). Therefore `{_}` is quite common in our scenario's. It would be nice if this would not require quoting at all (currently we have to resort to `{\\_}` ).

**Comment by vruizext** [comment]
I have a similar issue, I used a `TAG` fields to store URLs (don't know if it's the best choice), and then realized it's impossible to query the field by its value, because the URL contains some invalid characters. I would expect that non alphanumeric  characters are accepted, since the tag is stored verbatim, without any preprocessing. tokenization, stemming, ... , or isn't it?

**Comment by dvirsky** [comment]
@vruizext you can escape everything in a URL. However I would suggest using some hash of the URL to save memory.  But it works with proper escaping:

```
127.0.0.1:6379[1]> ft.create idx schema foo tag
OK
127.0.0.1:6379[1]> ft.add idx 1 1 fields foo "https://github.com/RedisLabsModules/RediSearch/issues/259#issuecomment-359281834"
OK
127.0.0.1:6379[1]> ft.tagvals idx foo
1) "https://github.com/redislabsmodules/redisearch/issues/259#issuecomment-359281834"

127.0.0.1:6379[1]> ft.search idx "@foo:{ https\\:\\/\\/github\\.com\\/RedisLabsModules\\/RediSearch\\/issues\\/259\\#issuecomment\\-359281834}"
1) (integer) 1
2) "1"
3) 1) "foo"
   2) "https://github.com/RedisLabsModules/RediSearch/issues/259#issuecomment-359281834"
```

**Comment by vruizext** [comment]
@dvirsky actually I use a hash of the url as the doc ID for my index, so I don't really need to search directly on the url, so this isn't an issue.

I guess, if I create the field as `url TEXT NOSTEM NOINDEX`, it will save more memory as saving it as a tag, right?

**Comment by dvirsky** [comment]
You don't even have to mention it in the schema if you don't want to search
by it.

**Comment by vruizext** [comment]
I'm including the url in the schema because I want it to be stored in the redis server, so I can recover it when I fetch the document later.

Is this command saving the url to the document hash, even the field is not in the schema?

`FT.ADD index doc_id FIELDS name foo url http://foobar.com`

**Comment by dvirsky** [comment]
yep. the schema only describes the index, but the documents themeselves are
schemaless.
The only reason NOINDEX exists is in conjunction with SORTABLE. If a field
is not sortable and not indexed there's no point in putting it in the
schema.

**Comment by vruizext** [comment]
Alright, thanks for clarifying it. That's great, it would have made my life easier if I would had known some months ago :) I assumed wrongly that redis was storing only the fields created in the schema.

**Comment by tw-bert** [comment]
Our team is, next to Lua, now also using Marc Gravell's .NET SE.RediSearch module.
Just to let you know this issue (support for quoted values) is still very relevant.
In Lua we worked around it by using the following code (which we will port to .NET as well, to prevent the same issues).
As soon as issue 259 is resolved, we can phase out those workarounds (which will fall over if RediSearch escape requirements change in the future).

```lua
local function FtEscapePunctuation(cRet)
  -- Escape RediSearch FT (full text) search punctuation characters, like '-' (becomes '\-')
  -- Also escape spaces, see: http://redisearch.io/Tags/
  -- For punctuation characters, see: https://github.com/RedisLabsModules/RediSearch/blob/master/src/toksep.h
  -- From the C code:
  --[[  [' '] = 1, ['\t'] = 1, [','] = 1, ['.'] = 1, ['/'] = 1, ['('] = 1, [')'] = 1,
        ['{'] = 1, ['}'] = 1,  ['['] = 1, [']'] = 1, [':'] = 1, [';'] = 1, ['\\'] = 1,
        ['~'] = 1, ['!'] = 1,  ['@'] = 1, ['#'] = 1, ['$'] = 1, ['%'] = 1, ['^'] = 1,
        ['&'] = 1, ['*'] = 1,  ['-'] = 1, ['='] = 1, ['+'] = 1, ['|'] = 1, ['\''] = 1,
        ['`'] = 1, ['"'] = 1,  ['<'] = 1, ['>'] = 1, ['?'] = 1,
    ]]
  -- Lua gsub: The lua magic characters are ( ) . % + - * ? [ ^ $
  -- So: prepend with '%' in the gsub pattern string (first parameter)
  return (cRet:gsub('[ \t,%./%(%){}%[%]:;\\~!@#%$%%%^&%*%-=%+|\'`"<>%?_%z]',{
      [' ' ]='\\ '  ,
      ['\t']='\\\t' ,
      [',' ]='\\,'  ,
      ['.' ]='\\.'  ,
      ['/' ]='\\/'  ,
      ['(' ]='\\('  ,
      [')' ]='\\)'  ,
      ['{' ]='\\{'  ,
      ['}' ]='\\}'  ,
      ['[' ]='\\['  ,
      [']' ]='\\]'  ,
      [':' ]='\\:'  ,
      [';' ]='\\;'  ,
      ['\\']='\\\\' ,
      ['~' ]='\\~'  ,
      ['!' ]='\\!'  ,
      ['@' ]='\\@'  ,
      ['#' ]='\\#'  ,
      ['$' ]='\\$'  ,
      ['%' ]='\\%'  ,
      ['^' ]='\\^'  ,
      ['&' ]='\\&'  ,
      ['*' ]='\\*'  ,
      ['-' ]='\\-'  ,
      ['=' ]='\\='  ,
      ['+' ]='\\+'  ,
      ['|' ]='\\|'  ,
      ['\'']='\\\'' ,
      ['`' ]='\\`'  ,
      ['"' ]='\\"'  ,
      ['<' ]='\\<'  ,
      ['>' ]='\\>'  ,
      ['?' ]='\\?'  ,
      -- Add underscore as well, seems needed
      ['_' ]='\\_'  ,
      -- Null char for specific situations and debugging
      [string.char(0x00)]='\\x00',
      }))
end
```

**Comment by mluggy** [comment]
@dvirsky solution works for urls but not em@ils. see this:

```
127.0.0.1:6379> ft.create user schema email TAG url TAG
OK
127.0.0.1:6379> ft.add user 1 1 fields email "michael@lugassy.net" url "https://lugassy.net"
OK
127.0.0.1:6379> ft.search user "@email:{michael\\@lugassy.net}"
1) (integer) 0
127.0.0.1:6379> ft.search user "@url:{https\\:\\/\\/lugassy\\.net}"
1) (integer) 1
2) "1"
3) 1) email
   2) "michael@lugassy.net"
   3) url
   4) "https://lugassy.net"
```

**Comment by garygreen** [comment]
Showing my interest in this as well. Currently query injection in Redisearch is a real concern for us.

Escaping queries to prevent injection is super complex and requires intricate knowledge of all the special characters in Redisearch - which differ in tag handling as well. It's currently so complex that we have considered escaping every character, for peace of mind.

With this idea it should substantially reduce attack vectors and simplify escaping, as presumably it would only require escaping of `\"'` ?

I hope this idea of quoted strings is prioritised, as it's a critical security consideration.

ping / @gkorland

**Comment by gkorland** [comment]
Please notice the way to avoid query injection is by using `PARAMS`.
See: https://redis.io/commands/ft.search/
